internal/venice-client-common/src/main/java/com/linkedin/venice/compression/GzipPool.java

```diff
@@ -0,0 +1,51 @@
+package com.linkedin.venice.compression;
+
+import com.linkedin.venice.utils.concurrent.CloseableThreadLocal;
+import java.io.IOException;
+import java.io.OutputStream;
+
+
+abstract class GzipPool {
+  private static class ReusableObjects implements AutoCloseable {
+    final SettableOutputStream stream = new SettableOutputStream();
+    final ReusableGzipOutputStream gzipOutputStream = new ReusableGzipOutputStream(stream);
+
+    @Override
+    public void close() throws Exception {
+      stream.close();
+    }
+  }
+
+  private static final CloseableThreadLocal<ReusableObjects> reusableObjectsThreadLocal =
+      new CloseableThreadLocal(ReusableObjects::new);
+
+  /**
+   * Retrieves an {@link ReusableGzipOutputStream} for the given {@link OutputStream}. Instances are pooled per thread.
+   *
+   * @param target
+   * @return
+   */
+  public static ReusableGzipOutputStream forStream(OutputStream target) {
+    reusableObjectsThreadLocal.get().stream.target = target;
+    return reusableObjectsThreadLocal.get().gzipOutputStream;
```

Comment:
Nitpick: What I meant in my comment about this is: can we do a single thread local access, store the `ReusableObject` in an in-line variable, and do both interactions with it there. This is the same pattern we have in many other parts of the code whenever there is more than one thing that needs to be thread local.

Reply:
I see, I thought threadLocal.get() is cheap.

Reply:
Nothing is cheap nowadays with inflation 🤣